src/Storages/ExternalStream/Kafka/KafkaSource.h

```diff
@@ -87,8 +87,10 @@ class KafkaSource final : public ISource
     Int32 record_consume_timeout_ms = 100;
 
     /// For checkpoint
-    /// FIXME, switch to llvm-15
-    std::atomic<CheckpointContext *> ckpt_ctx;
+    /// FIXME, switch to llvm-15 atomic shared_ptr
```

Comment:
We can consider refactor it out as a util struct like below and reuse everywhere ?

```
/// \CheckpointRequest applies double check locking to minimize using mutex 
/// directly for better performance
class CheckpointRequest
{
public:
    CheckpointContextPtr poll() noexcept 
    { 
        CheckpointContextPtr new_request;
        if (has_ckpt_request)
        {
            std::scoped_lock lock(ckpt_mutex);
            current.swap(ckpt_ctx);
            has_ckpt_request = false;
        }
        return new_request;
    }

     void setCheckpointRequestCtx(CheckpointContextPtr ckpt_ctx_)
     {
         assert(!has_ckpt_request);
         std::scoped_lock lock(ckpt_mutex);
         assert(!ckpt_ctx);
         ckpt_ctx.swap(ckpt_ctx_);
         has_ckpt_request = true;
     }

private:
    std::atomic<bool> has_ckpt_request = false;
    std::mutex ckpt_mutex;
    CheckpointContextPtr ckpt_ctx TSA_PT_GUARDED_BY(ckpt_mutex);
};
```